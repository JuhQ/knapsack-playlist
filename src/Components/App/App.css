@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 2s linear;
    display: inline-block;
    font-size: 2rem;
  }
}

h1,
.security {
  background-image: linear-gradient(
    69deg,
    var(--color-three),
    var(--color-four),
    var(--color-two),
    var(--color-one)
  );
  background-clip: text;
  -webkit-text-fill-color: transparent;
  position: relative;
  transform: all 0.3s ease;
  display: inline-block;
  margin: 2rem;
}

h1::after {
  text-shadow: 0px 0px 100px cyan;
  content: attr(data-text);
  position: absolute;
  left: 0;
  top: 0;
  z-index: -1;
  transition: all 2s cubic-bezier(0.6, -0.28, 0.735, 0.045);
}

h1:hover::after {
  text-shadow: 0px 0px 100px rgb(250, 104, 255);
}

.security {
  -webkit-text-security: square;
  transition: all 0.2s cubic-bezier(0.6, 0.04, 0.98, 0.335);
}

.security:hover {
  -webkit-text-security: square;
  font-size: 2rem;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
