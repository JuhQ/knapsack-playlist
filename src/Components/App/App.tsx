/* eslint-disable react/require-default-props */
import "./App.css";

import React, { useEffect, useState } from "react";
import { Container, Grid } from "semantic-ui-react";

import knapsack from "../../Algorithms/knapsack/knapsack";
import YoutubeMusic from "../../Models/Youtube/Youtube";
import { YoutubeItem } from "../../types";
import sample from "../../Utils/utils";
import EntertainmentSystem from "../EntertainmentSystem/EntertainmentSystem";
import Generator from "../Generator/Generator";

interface Props {
  initialSetSize?: number;
  initialPlaylistLength?: number;
  initialCreating?: boolean;
  initialPlaylist?: YoutubeItem[] | null;
  initialGeneratorSubmitted?: boolean;
}

const App: React.FC<Props> = ({
  initialSetSize = 0,
  initialPlaylistLength = 0,
  initialCreating = false,
  initialPlaylist = null,
  initialGeneratorSubmitted = false,
}: Props) => {
  const [dataSetSize, setDataSetSize] = useState<number>(initialSetSize);
  const [playlistLength, setPlaylistLength] = useState<number>(
    initialPlaylistLength
  );
  const [playlist, setPlaylist] = useState<YoutubeItem[] | null>(
    initialPlaylist
  );
  const [creating, setCreating] = useState<boolean>(initialCreating);
  const [generatorSubmitted, setGeneratorSubmitted] = useState<boolean>(
    initialGeneratorSubmitted
  );

  useEffect(() => {
    if (generatorSubmitted) {
      setCreating(true);
      setTimeout(() => {
        setPlaylist(
          knapsack(sample(YoutubeMusic(), dataSetSize), playlistLength)
        );
        setCreating(false);
        setGeneratorSubmitted(false);
      }, 0);
    }
  }, [dataSetSize, playlistLength, generatorSubmitted]);

  return (
    <Grid>
      <Grid.Column width={16}>
        <Container textAlign="center">
          <h1 data-text="Playlist Generator">Playlist Generator</h1>
          <span className="App-logo" role="img" aria-label="rock on!">
            ðŸŽ¸
          </span>
          <span className="App-logo" role="img" aria-label="rock on!">
            ðŸ¤˜
          </span>
        </Container>

        <Generator
          onSubmit={(value) => {
            setDataSetSize(value.dataSetSize);
            setPlaylistLength(value.playlistLength);
            setGeneratorSubmitted(true);
          }}
        />
      </Grid.Column>
      <Grid.Column width={16}>
        <Container>
          {dataSetSize === 0 && <>Select dataset size</>}

          {(dataSetSize > 0 || playlistLength > 0) &&
            !creating &&
            !playlist?.length && (
              <>Playlist generated but no songs were found</>
            )}
          {creating && (
            <>
              Generating playlist. Depending on input size and desired playlist
              length, this might take a while.
            </>
          )}
        </Container>
      </Grid.Column>
      <Grid.Column width={16}>
        {dataSetSize > 0 && playlist?.length && (
          <EntertainmentSystem list={playlist} length={playlistLength} />
        )}
        <span className="security">Knapsack generated playlists rule!</span>
      </Grid.Column>
    </Grid>
  );
};
export default App;
