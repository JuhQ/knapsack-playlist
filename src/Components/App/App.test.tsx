import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import React from "react";
import { act, create } from "react-test-renderer";

import App from "./App";

describe("App", () => {
  it("should render", () => {
    const component = create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should submit empty dataset size and playlist length", () => {
    const wrapper = shallow(<App />);
    wrapper
      .find("Generator")
      .simulate("submit", { dataSetSize: 0, playlistLength: 0 });

    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it("should submit only length", () => {
    const wrapper = shallow(<App />);
    wrapper
      .find("Generator")
      .simulate("submit", { dataSetSize: 0, playlistLength: 10 });

    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it("should submit only dataset size", () => {
    const wrapper = shallow(<App />);
    wrapper
      .find("Generator")
      .simulate("submit", { dataSetSize: 10, playlistLength: 0 });

    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it("should submit proper dataset size and playlist length", () => {
    const wrapper = shallow(<App />);
    wrapper
      .find("Generator")
      .simulate("submit", { dataSetSize: 5, playlistLength: 3600 });

    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe("initial values", () => {
    it("should submit only length", () => {
      const wrapper = shallow(
        <App
          initialPlaylistLength={100}
          initialSetSize={100}
          initialPlaylist={[{ id: "1", title: "test", seconds: 2 }]}
          initialGeneratorSubmitted
          initialCreating
        />
      );
      wrapper
        .find("Generator")
        .simulate("submit", { dataSetSize: 0, playlistLength: 10 });

      act(() => {
        expect(shallowToJson(wrapper)).toMatchSnapshot();
      });
    });

    it("should submit only dataset size", () => {
      const wrapper = shallow(
        <App
          initialPlaylistLength={100}
          initialSetSize={100}
          initialPlaylist={[{ id: "1", title: "test", seconds: 2 }]}
          initialGeneratorSubmitted
          initialCreating
        />
      );
      wrapper
        .find("Generator")
        .simulate("submit", { dataSetSize: 10, playlistLength: 0 });

      act(() => {
        expect(shallowToJson(wrapper)).toMatchSnapshot();
      });
    });

    it("should submit proper dataset size and playlist length", () => {
      const wrapper = shallow(
        <App
          initialPlaylistLength={100}
          initialSetSize={100}
          initialPlaylist={[{ id: "1", title: "test", seconds: 2 }]}
          initialGeneratorSubmitted
          initialCreating
        />
      );
      wrapper.find("Generator").simulate("submit", {
        value: { dataSetSize: 5, playlistLength: 3600 },
      });

      act(() => {
        expect(shallowToJson(wrapper)).toMatchSnapshot();
      });
    });
  });
});
