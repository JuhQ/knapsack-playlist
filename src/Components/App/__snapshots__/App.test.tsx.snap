// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App initial values should submit only dataset size 1`] = `
<Grid>
  <GridColumn
    width={16}
  >
    <Container
      textAlign="center"
    >
      <h1
        data-text="Playlist Generator"
      >
        Playlist Generator
      </h1>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸŽ¸
      </span>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸ¤˜
      </span>
    </Container>
    <Generator
      onSubmit={[Function]}
    />
  </GridColumn>
  <GridColumn
    width={16}
  >
    <Container>
      Generating playlist. Depending on input size and desired playlist length, this might take a while.
    </Container>
  </GridColumn>
  <GridColumn
    width={16}
  >
    <EntertainmentSystem
      length={0}
      list={
        Array [
          Object {
            "id": "1",
            "seconds": 2,
            "title": "test",
          },
        ]
      }
    />
    <span
      className="security"
    >
      Knapsack generated playlists rule!
    </span>
  </GridColumn>
</Grid>
`;

exports[`App initial values should submit only length 1`] = `
<Grid>
  <GridColumn
    width={16}
  >
    <Container
      textAlign="center"
    >
      <h1
        data-text="Playlist Generator"
      >
        Playlist Generator
      </h1>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸŽ¸
      </span>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸ¤˜
      </span>
    </Container>
    <Generator
      onSubmit={[Function]}
    />
  </GridColumn>
  <GridColumn
    width={16}
  >
    <Container>
      Select dataset size
      Generating playlist. Depending on input size and desired playlist length, this might take a while.
    </Container>
  </GridColumn>
  <GridColumn
    width={16}
  >
    <span
      className="security"
    >
      Knapsack generated playlists rule!
    </span>
  </GridColumn>
</Grid>
`;

exports[`App initial values should submit proper dataset size and playlist length 1`] = `
<Grid>
  <GridColumn
    width={16}
  >
    <Container
      textAlign="center"
    >
      <h1
        data-text="Playlist Generator"
      >
        Playlist Generator
      </h1>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸŽ¸
      </span>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸ¤˜
      </span>
    </Container>
    <Generator
      onSubmit={[Function]}
    />
  </GridColumn>
  <GridColumn
    width={16}
  >
    <Container>
      Generating playlist. Depending on input size and desired playlist length, this might take a while.
    </Container>
  </GridColumn>
  <GridColumn
    width={16}
  >
    <span
      className="security"
    >
      Knapsack generated playlists rule!
    </span>
  </GridColumn>
</Grid>
`;

exports[`App should render 1`] = `
<div
  className="ui grid"
>
  <div
    className="sixteen wide column"
  >
    <div
      className="ui center aligned container"
    >
      <h1
        data-text="Playlist Generator"
      >
        Playlist Generator
      </h1>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸŽ¸
      </span>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸ¤˜
      </span>
    </div>
    <div
      className="ui container"
    >
      <form
        className="ui form"
        onSubmit={[Function]}
      >
        <div
          className="ui bottom aligned grid"
        >
          <div
            className="six wide computer sixteen wide mobile column"
          >
            Size of data set:
            <br />
            <small>
              This is the amount of random items to be selected from the database for the knapsack algorithm
            </small>
            <br />
            <div
              className="ui large right labeled input"
            >
              <input
                onChange={[Function]}
                placeholder="Enter desired data set size"
                type="number"
                value={20}
              />
              <div
                className="ui basic label label"
                onClick={[Function]}
              >
                videos
              </div>
            </div>
          </div>
          <div
            className="ten wide computer sixteen wide mobile column"
          >
            <button
              className="preset dataset"
              onClick={[Function]}
              type="button"
            >
              20
            </button>
            <button
              className="preset dataset"
              onClick={[Function]}
              type="button"
            >
              25
            </button>
            <button
              className="preset dataset"
              onClick={[Function]}
              type="button"
            >
              30
            </button>
            <button
              className="preset dataset"
              onClick={[Function]}
              type="button"
            >
              35
            </button>
            <button
              className="preset dataset"
              onClick={[Function]}
              type="button"
            >
              40
            </button>
            <button
              className="preset dataset"
              onClick={[Function]}
              type="button"
            >
              50
            </button>
          </div>
        </div>
        <div
          className="ui bottom aligned grid"
        >
          <div
            className="six wide computer sixteen wide mobile column"
          >
            Length of playlist:
            <br />
            <div
              className="ui large right labeled input"
            >
              <input
                onChange={[Function]}
                placeholder="Enter desired playlist length"
                type="number"
                value={3600}
              />
              <div
                className="ui basic label label"
                onClick={[Function]}
              >
                seconds
              </div>
            </div>
          </div>
          <div
            className="ten wide computer sixteen wide mobile column"
          >
            <button
              className="preset length"
              onClick={[Function]}
              type="button"
            >
              300
            </button>
            <button
              className="preset length"
              onClick={[Function]}
              type="button"
            >
              600
            </button>
            <button
              className="preset length"
              onClick={[Function]}
              type="button"
            >
              1800
            </button>
            <button
              className="preset length"
              onClick={[Function]}
              type="button"
            >
              3600
            </button>
            <button
              className="preset length"
              onClick={[Function]}
              type="button"
            >
              7200
            </button>
          </div>
        </div>
        <button
          className="submit"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
  <div
    className="sixteen wide column"
  >
    <div
      className="ui container"
    >
      Select dataset size
    </div>
  </div>
  <div
    className="sixteen wide column"
  >
    <span
      className="security"
    >
      Knapsack generated playlists rule!
    </span>
  </div>
</div>
`;

exports[`App should submit empty dataset size and playlist length 1`] = `
<Grid>
  <GridColumn
    width={16}
  >
    <Container
      textAlign="center"
    >
      <h1
        data-text="Playlist Generator"
      >
        Playlist Generator
      </h1>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸŽ¸
      </span>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸ¤˜
      </span>
    </Container>
    <Generator
      onSubmit={[Function]}
    />
  </GridColumn>
  <GridColumn
    width={16}
  >
    <Container>
      Select dataset size
    </Container>
  </GridColumn>
  <GridColumn
    width={16}
  >
    <span
      className="security"
    >
      Knapsack generated playlists rule!
    </span>
  </GridColumn>
</Grid>
`;

exports[`App should submit only dataset size 1`] = `
<Grid>
  <GridColumn
    width={16}
  >
    <Container
      textAlign="center"
    >
      <h1
        data-text="Playlist Generator"
      >
        Playlist Generator
      </h1>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸŽ¸
      </span>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸ¤˜
      </span>
    </Container>
    <Generator
      onSubmit={[Function]}
    />
  </GridColumn>
  <GridColumn
    width={16}
  >
    <Container>
      Playlist generated but no songs were found
    </Container>
  </GridColumn>
  <GridColumn
    width={16}
  >
    <span
      className="security"
    >
      Knapsack generated playlists rule!
    </span>
  </GridColumn>
</Grid>
`;

exports[`App should submit only length 1`] = `
<Grid>
  <GridColumn
    width={16}
  >
    <Container
      textAlign="center"
    >
      <h1
        data-text="Playlist Generator"
      >
        Playlist Generator
      </h1>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸŽ¸
      </span>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸ¤˜
      </span>
    </Container>
    <Generator
      onSubmit={[Function]}
    />
  </GridColumn>
  <GridColumn
    width={16}
  >
    <Container>
      Select dataset size
      Playlist generated but no songs were found
    </Container>
  </GridColumn>
  <GridColumn
    width={16}
  >
    <span
      className="security"
    >
      Knapsack generated playlists rule!
    </span>
  </GridColumn>
</Grid>
`;

exports[`App should submit proper dataset size and playlist length 1`] = `
<Grid>
  <GridColumn
    width={16}
  >
    <Container
      textAlign="center"
    >
      <h1
        data-text="Playlist Generator"
      >
        Playlist Generator
      </h1>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸŽ¸
      </span>
      <span
        aria-label="rock on!"
        className="App-logo"
        role="img"
      >
        ðŸ¤˜
      </span>
    </Container>
    <Generator
      onSubmit={[Function]}
    />
  </GridColumn>
  <GridColumn
    width={16}
  >
    <Container>
      Playlist generated but no songs were found
    </Container>
  </GridColumn>
  <GridColumn
    width={16}
  >
    <span
      className="security"
    >
      Knapsack generated playlists rule!
    </span>
  </GridColumn>
</Grid>
`;
